### PR TITLE
correction due to method renamed in promoted-builds-plugin

### PR DESCRIPTION
**Pull request by froque**
added unit test for normal and promotion jobs

See:
https://issues.jenkins-ci.org/browse/JENKINS-59704
https://issues.jenkins-ci.org/browse/JENKINS-59600
https://issues.jenkins-ci.org/browse/JENKINS-58337
https://github.com/jenkinsci/promoted-builds-plugin/pull/140

### CODE REVIEW

**Comment by JoaoPPinto** [comment]
@slide any chance this will get looked at in the near future?

**Comment by slide** [comment]
I don't maintain this plugin anymore, it is up for adoption

**Comment by froque** [comment]
@slide I am willing to be a maintainer if that is OK with you

**Comment by slide** [comment]
Are you willing to maintain po-ftp, po-cifs and po-ssh as well?

**Comment by oleg-nenashev** [comment]
https://github.com/jenkinsci/promoted-builds-plugin/pull/144 for the fix on the promoted builds side

**Comment by gmcdonald** [comment]
I am wondering, I could be reading it wrong, but should this fix now be updated to match getTargetBuildOrFail(); ?

**Comment by gmcdonald** [comment]
Would love to merge this. but still feel seeing as promoted builds plugin upstream has changed to getTargetBuildOrFail();  that this should be altered to match.

**Comment by froque** [comment]
@gmcdonald changed to `getTargetBuildOrFail` 

promoted-builds has now removed `getTarget` in release [867.v7c3a_b_83a_eb_79](https://github.com/jenkinsci/promoted-builds-plugin/releases/tag/867.v7c3a_b_83a_eb_79) and Jenkins now fails with `The base directory does not exist. If this is a promotion, have you "Archived the artifacts"?`. 

So this also fixes https://github.com/jenkinsci/publish-over-ssh-plugin/issues/85 

So much history to a one line change. I had to recapitulate all the changes to go back to this pull request:
| date  | change link | component | description | JIRA |  
| ------------- | ------------- | ------------- | ------------- | ------------- |
| 2018-09-27  | https://github.com/jenkinsci/jenkins/commit/ba33bd67cdaef87aba8a4e95dca8dcf108a7d73f  | jenkins 2.138.2  |  adds interface StaplerProxy to class Run, with getTarget() method  |   |
| 2019-06-01 | https://github.com/jenkinsci/promoted-builds-plugin/commit/0a9e67654a0f072601fb3a24d420c3c09ebfd843 https://github.com/jenkinsci/promoted-builds-plugin/pull/124 | promoted-builds-3.3  | upgrade jenkins.version property from 2.60.3 to 2.138.4  | [JENKINS-59600](https://issues.jenkins.io/browse/JENKINS-59600) [JENKINS-58337](https://issues.jenkins.io/browse/JENKINS-58337) [JENKINS-58534](https://issues.jenkins.io/browse/JENKINS-58534) |
| 2019-10-07  |  https://github.com/jenkinsci/promoted-builds-plugin/commit/cf1128c914f2d84260c202ce58f6e559384761c6 https://github.com/jenkinsci/promoted-builds-plugin/pull/140  | promoted-builds-3.4  | getTarget renamed getTargetBuild  | [JENKINS-59704](https://issues.jenkins-ci.org/browse/JENKINS-59704) |
| 2019-10-15 | https://github.com/jenkinsci/git-plugin/commit/54d4a89246c8b8b274ca93b24f1438852599a261 https://github.com/jenkinsci/git-plugin/pull/774 | git |   Bump promoted-builds from 3.3 to 3.4  | | 
| 2019-10-17 | https://github.com/jenkinsci/git-plugin/commit/91ef2565d4a4fe2669430400cf52025c44985e25 https://github.com/jenkinsci/git-plugin/pull/775 | git |  Revert "Depend on promoted builds 3.4" | |
| 2019-10-17 | https://github.com/jenkinsci/git-plugin/commit/a69d8bc6ed3e375880b285316f1d4f602b22bc0a https://github.com/jenkinsci/git-plugin/pull/777 | git | Use promoted-builds 3.2, not 3.3 or 3.4 | |
| 2019-10-17 | https://github.com/jenkinsci/parameterized-trigger-plugin/pull/130 | parameterized-trigger | Update dependency on the Jenkins core and the Promoted Builds plugin | |
| 2019-12-02 | https://github.com/jenkinsci/promoted-builds-plugin/pull/144 | promoted-builds-3.5 | adds back getTarget and adds new getTargetBuildOrFail | |
| *open* | https://github.com/jenkinsci/publish-over-plugin/pull/10 https://github.com/jenkinsci/publish-over-plugin/pull/13 | publish-over-x | correction due to method renamed in promoted-builds-plugin | |
| 2022-02-08 | https://github.com/jenkinsci/promoted-builds-plugin/pull/159 | promoted-builds | Remove Promotion#getTarget | [JENKINS-67045](https://issues.jenkins.io/browse/JENKINS-67045) |

File: src/main/java/jenkins/plugins/publish_over/BPPlugin.java [hunk 1]
```diff
@@ -107,7 +107,7 @@ public void perform(final Run<?, ?> build, final FilePath workspace, final Launc
         if (PROMOTION_CLASS_NAME.equals(build.getClass().getCanonicalName())) {
             AbstractBuild<?, ?> promoted;
             try {
-                final Method getTarget = build.getClass().getMethod("getTarget", (Class<?>[]) null);
+                final Method getTarget = build.getClass().getMethod("getTargetBuildOrFail", (Class<?>[]) null);
                 promoted = (AbstractBuild) getTarget.invoke(build, (Object[]) null);
```
Comment (slide) [inline src/main/java/jenkins/plugins/publish_over/BPPlugin.java:111]: Something along these lines perhaps?

```suggestion
                final Method getTarget = build.getClass().getMethod("getTargetBuildOrFail", (Class<?>[])null);
                if(getTarget == null) {
                    getTarget = build.getClass().getMethod("getTarget", (Class<?>[])null);
                }
                if(getTarget != null) {
                    promoted = (AbstractBuild) getTarget.invoke(build, (Object[]) null);
                }
```

Comment (slide) [reply src/main/java/jenkins/plugins/publish_over/BPPlugin.java:111]: Don't commit this suggestion, it wouldn't let me include the modified line.

**Comment by froque** [comment]
> Wouldn't you need to check for both method's existing (with a preference for the newer method)? Some people may not have the newer version of promoted builds plugin, so the getBuild method would still exist.

That would only be needed when running with promoted-builds 3.4 (2019) or lower.  

I tried to run a newer Jenkins with promoted-builds 3.4 and is not usable. Could not save FreeStyleJobs due to javascript errors https://issues.jenkins.io/browse/JENKINS-65304

I don't think it should support something that old.

**Comment by gmcdonald** [comment]
Thanks @froque for the updated fixes, apologies for not testing until now. 

I am getting an error with 'mvn test' and also 'mvn clean -ntp verify'

```[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.477 s
[INFO] Finished at: 2023-07-08T20:51:59+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:testCompile (default-testCompile) on project publish-over: Compilation failure
[ERROR] /Users/gmcdonald/jenkins/publish-over-plugin/src/test/java/jenkins/plugins/publish_over/BPPluginTest.java:[3,37] package com.gargoylesoftware.htmlunit does not exist
[ERROR] 
```

**Comment by froque** [comment]
@gmcdonald rebased and resolved

**Comment by gmcdonald** [comment]
Perfect, thanks! - Merging

**Comment by drvora** [comment]
@froque @gmcdonald [https://github.com/jenkinsci/publish-over-ssh-plugin/issues/85] is still not resolved. Following combination still throws `The base directory does not exist. If this is a promotion, have you "Archived the artifacts"?`.

Publish-over-ssh -- [1.25](https://updates.jenkins.io/download/plugins/publish-over-ssh/1.25/publish-over-ssh.hpi)
Promoted-build -- [892.vd6219fc0a_efb](https://updates.jenkins.io/download/plugins/promoted-builds/892.vd6219fc0a_efb/promoted-builds.hpi)

**Comment by froque** [comment]
A new version of this plugin should be created with the fix. 

The latest release is 0.22 from 5 years ago

@gmcdonald can you create a new release ?

**Pull request merged**